Control speed of scene when playing some animation

sparshg opened the pull request:
<!-- Thank you for contributing to Manim! Learn more about the process in our contributing guidelines: https://docs.manim.community/en/latest/contributing.html -->

## Overview: What does this pull request change?
<!-- If there is more information than the PR title that should be added to our release changelog, add it in the following changelog section. This is optional, but recommended for larger pull requests. -->
<!--changelog-start-->
Added speed control
<!--changelog-end-->

## Motivation and Explanation: Why and how do your changes improve the library?
<!-- Optional for bugfixes, small enhancements, and documentation-related PRs. Otherwise, please give a short reasoning for your changes. -->

### Basic idea / How it works

This PR changes the time stamps generated by `time_progression` to change the overall speed to the scene.
For example the `time_progression` generated for`run_time=1` animation without changing speed is/was `[0, dt, 2dt, 3dt....1)`, so the difference between each term (which = speed of scene) is `dt` in this case

Now, if we change speed of scene to half till we reach the end of this `run_time=1` animation, the time stamps on `time_progression` is made to follow a parabolic curve, where the initial slope of curve (initial speed) is `dt` and final is `dt/2`. Means something close to `[0, dt, ....diff b/w terms lessens... 1-dt/2, 1)` should be generated.

So to get this I generated time stamps normally, `[0, dt, 2dt, 3dt..... k dt)` (where `f(k dt) = 1`) and then pass them through such a quadratic function where initial slope = `dt` and final slope = `dt/2` and we get modified time stamps between `[0, 1)`.

**The number of time stamps/frames as you can see here will be more than run_time** (`k * run_time`)

After that added the nodes which gives further control.
Added `relative_speed_mode` parameter now to strictly respect `run_time` but the speed changes are only relative to each other in the given animation

### Usage

If you want to change speed of an animation, say when its 1/4th done we want its speed to be at 0.1x and when its half done the speed should go to 0.8x. 

#### Syntax 1:
```py
with self.scene_speed_changing({0.25: 0.1, 0.5: 0.8}, retain_final_speed=False, relative_speed_mode=False):
    self.play(your_animation, another_anim_maybe, run_time=2)
self.play(after_anim)
```

#### Syntax 2:
```py
self.play(
    your_animation,
    another_anim_maybe,
    run_time=2,
    animation_speedinfo={0.25: 0.1, 0.5: 0.8},
    retain_final_speed=False,
    relative_speed_mode=False
)
self.play(after_anim)
```
The total time for animation will be greater than `run_time=2` in this case...
`after_anim` will also be played at 0.8x because the speed change is retained by default,

Another example if you want to play multiple animations but not simultaneously and spread the speed effect over all the animations, you can combine them using `AnimationGroup`
```py
with self.scene_speed_changing({1: 0.1}, False):
    self.play(
        AnimationGroup(
            s.animate.shift(RIGHT * 5),
            FadeIn(s)
            c.animate.shift(RIGHT * 5),
            rate_func=linear,
            lag_ratio=1,
        )
    )
```
### Example

https://user-images.githubusercontent.com/43041139/133597529-8563a00e-8900-4cbc-9d58-064270172230.mp4

```py
...
        apply_force_to_balls()
        with self.scene_speed_changing({0.2: 0.1, 0.8: 0.08, 1: 1}):
            self.play(
                self.camera.frame.animate.scale(1.3).shift(0.3 * config.frame_y_radius * UP),
                rate_func=rush_from,
                run_time=0.4,
            )
        self.wait()
```
## Links to added or changed documentation pages
<!-- Please add links to the affected documentation pages (edit the description after opening the PR). The link to the documentation for your PR is https://manimce--####.org.readthedocs.build/en/####/, where #### represents the PR number. -->


## Further Information and Comments
<!-- If applicable, put further comments for the reviewers here. -->



<!-- Thank you again for contributing! Do not modify the lines below, they are for reviewers. -->
## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

behackl commented on manim/scene/scene.py:
```diff
@@ -849,8 +883,36 @@ def get_time_progression(
+                times = np.array([])
+                on_node = 0
+
+                for dur, s in zip(np.diff(nodes), speeds):
```
The code in this loop is rather hard to read. Could you please add some comments and more explicit variable names? (`dur` -> `duration`, `s` -> `speed`, etc.; `dt` is fine.)

behackl commented on manim/scene/scene.py:
```diff
@@ -849,8 +883,36 @@ def get_time_progression(
+                for dur, s in zip(np.diff(nodes), speeds):
+                    rel_speed_change = s / self.speed
+                    adj_func = np.vectorize(
+                        lambda x: (rel_speed_change ** 2 - 1) / 4 * x ** 2 + x,
```
What does this function do, precisely? What if users want to use a custom function here?

behackl commented on manim/scene/scene.py:
```diff
@@ -763,7 +793,7 @@ def compile_animations(self, *args, **kwargs):
         return animations
 
-    def _get_animation_time_progression(self, animations, duration):
+    def _get_animation_time_progression(self, animations, duration, speedinfo):
```
It would be good to replace `speedinfo` (in general, not only here) with something more explicit such that the naming of the newly introduced attributes is a bit more uniform. How about `animation_speed`?

behackl commented on manim/scene/scene.py:
```diff
@@ -849,8 +883,36 @@ def get_time_progression(
+                    adj_func = np.vectorize(
+                        lambda x: (rel_speed_change ** 2 - 1) / 4 * x ** 2 + x,
+                    )
+                    func_unity_at = 2 / (rel_speed_change + 1)
```
I don't see what this value is or where it comes from.

behackl commented on manim/scene/scene.py:
```diff
@@ -849,8 +883,36 @@ def get_time_progression(
+                        + nodes[on_node] * run_time,
+                    )
+                    self.speed = s
+                    on_node += 1
```
I get the feeling that it would be cleaner to start with some `previous_node = 0`, then iterating over the dictionary via something like `for node, speed in sorted(speedinfo.items())` and computing the duration as `duration = node - previous_node`, while at the end of the loop setting `previous_node` to `node`.

behackl reviewed (changes requested):
Sorry it took so long for someone to review this.

I very much like the idea here! Still, I think there are some problems with the current implementation that might cause frustration with users (or developers that need to maintain this feature).

Additionally to the comments I've left in the code itself, I am not sure about the following points:

- The fact that setting the animation speed "overrides" (or rather does not respect) the animation time is not great. Have you considered to rescale the timestamps so that, ultimately, `run_time` is still the actual run time of the animation?
- It should be possible for users to pass `speedinfo` (or whatever the parameter is named in the end) to `self.play`. And on a related note, I find it odd that `self.speedinfo` is passed to `self._get_animation_time_progression`, the method would have direct access to the attributes itself (but the same holds for `self.animations` and `self.duration`, so this is maybe rather a general issue with the code than with your implementation).

Obviously, there is a lot of my personal opinion included here; I'm certainly open for discussions and I won't block this PR over disagreeing on some details. (In that case, more feedback from other devs would be good though.)

Most importantly: thank you very much for your efforts!

sparshg commented:
Thanks for your review, this definitely needs more work and cleanup, will try fixing everything after few days.

behackl commented on manim/scene/scene.py:
```diff
@@ -727,6 +731,32 @@ def get_moving_and_static_mobjects(self, animations):
+    def changing_scene_speed_to(
+        self,
+        speedinfo: Dict[float, float],
+        retain_final_speed=True,
```
By default, the context manager should only change things inside of its context. Retaining the final speed by default goes against the idea of a context manager -- this could actually be made into an argument of whether a context manager is even the right tool for this job.

```suggestion
        retain_final_speed=False,
```

behackl commented:
> Now I just want to apply slow motion to second animation after I am done making my animations, (or speed up things like time lapse?)
> 
> ```python
> # self.speed is 1
> self.play(s.animate.shift(2 * RIGHT), rate_func=linear)
> 
> # changing self.speed from 1 to 0.1
> self.play(
>     s.animate.shift(2 * RIGHT),
>     rate_func=linear,
>     animation_speedinfo={1: 0.1},
> )
> ```

I am not sure I understand what you want to do here. If you wanted the second animation to be in slow motion, shouldn't you pass `animation_speedinfo={0: 0.1}`? I don't really see anything in the video either.

> But it generates additional frames (which is proportional to `run_time`), 55 in this case instead of 30 for second animation. Now you can adjust it to follow `run_time` but that would basically mean to decrease frames back to 30, but then the animation won't be the same...

What I am suggesting is the following:

First, the code generates the modified time stamps according to the passed `animation_speedinfo`. In my head, `Animation(..., animation_speedinfo={0: 0.1, 0.5: 0.5, 0.9: 2})` generates weights for the different parts of the animation; the first part gets a weight (which is their "unscaled time length") of (0.5 - 0) / 0.1 = 5, the second part a weight of (0.9 - 0.5) / 0.5 = 0.8, and the third part a weight of (1 - 0.9) / 2 = 0.05. But then, instead of simply building the animation with 5.85 sec runtime, I'd rescale the timestamps such that they match the given runtime. Yes, this might cause some weirdness like an explicitly set speed of 1 not always being equal to the unmodified animation speed of the scene, but in my opinion respecting the passed `run_time` of animations is more important than absolute speed measurements.

There could, of course, be an additional keyword argument that allows users to explicitly disable rescaling to match the run time so that their speed specifications hold globally. (This is probably the best way to go.)

behackl commented:
> It should be possible for users to pass speedinfo (or whatever the parameter is named in the end) to `Scene.play`. 

This comment from my previous review seems not to have been addressed.

sparshg commented:
Added parameters for `Scene.play`, and the `relative_speed_mode` so that when true, run_time is followed strictly, added doc strings too and also added check for occasional floating point error that didn't generate exact amount of frames expected, so now it should work correctly. I would fix if there's something else left, but it seems good to go imo

sparshg commented:
Closed in favor of #2749
